### DISCUSSION TITLE
BullTrading copytrading and bots

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#924] [Show and tell]
Looks like they’re doing some cool [signal processing](https://docs.bulltrading.be/bulltrading-designer/editor/create-your-first-indicator/lets-look-at-the-created-file/signal-processing) over at [BullTrading.be](https://bulltrading.be)
Nicely done!

**Comment by Brizze0001** [comment]
Hey Dave, 

Thank you for your library, I will try to publish my contributions later. If you want more info on the project, don't hesitate to ask your questions.

Have a nice day.

**Reply by DaveSkender** [reply]
Would love to hear your project goals, how our library is helping you, and how we can improve your experience.  I can also chat privately if you want to connect on Twitter DM [@daveskender](https://twitter.com/messages/compose?recipient_id=27475431).

**Reply by DaveSkender** [reply]
I also just created an open [2022 Retrospective](https://github.com/DaveSkender/Stock.Indicators/discussions/940) discussion for everyone.

**Reply by Brizze0001** [reply]
Hey,

With our team we are trying to build the most powerful strategy editor on the market. : https://bulltrading.be/

The editor allows you to build strategies by assembling blocks of indicators such as the crossing of moving averages, the RSI, MACD etc... Currently we have built around 20 default indicators, this list will tend to grow with time. Your creations can either be used by yourself or you can rent them to other users of the platform.

The editor wants to be innovative by its ability to accommodate code in LUA (https://www.lua.org/about.html) and to transpose this code in the form of blocks. You can therefore create your own indicators coded by yourself and even sell your indicator while keeping its private code.

We will then continue to work on the editor in order to allow deep-learning, the possibility of modifying certain parameters according to market volatility, etc.

Here is an example of a moving average crossover strategy : 
![image](https://user-images.githubusercontent.com/87341161/201448429-7fba1458-4f62-4295-a62a-47931fdb7273.png)


Look here the code of the moving average cross blocks : (All globals variable can be edited in the blocks)

![image](https://user-images.githubusercontent.com/87341161/201448903-61f796bf-3143-4577-85b3-b0c45b3d9d69.png)

![image](https://user-images.githubusercontent.com/87341161/201448931-02470104-cf91-404d-8c98-a87a0b6630ee.png)

![image](https://user-images.githubusercontent.com/87341161/201448944-3e034706-b11d-4644-8de7-0f74adeb5721.png)


```

PeriodSlowMovingAverage = 200
TypeSlowMovingAverage = {Simple_SMA = true, Exponential_EMA = false, Weighted_WMA = false, Smoothed_SMMA = false}

PeriodFastMovingAverage = 100
TypeFastMovingAverage = {Simple_SMA = true, Exponential_EMA = false, Weighted_WMA = false, Smoothed_SMMA = false}

SignalType = {CrossOver = false, Position = true}
TriggerType = {Bullish = true, Bearish = false}

function ComputeOutputs(quotes, quotesInFiveMinutes)
    local outputs, chartFMA, chartSMA, slowMA, fastMA  = {}, {}, {}, {}, {}

    if(TypeSlowMovingAverage.Simple_SMA == true) then
        slowMA = quotes.GetSma(PeriodSlowMovingAverage).ToResults()
    elseif(TypeSlowMovingAverage.Exponential_EMA == true) then
        slowMA = quotes.GetEma(PeriodSlowMovingAverage).ToResults()
    elseif(TypeSlowMovingAverage.Weighted_WMA == true) then
        slowMA = quotes.GetWma(PeriodSlowMovingAverage).ToResults()
    elseif(TypeSlowMovingAverage.Smoothed_SMMA == true) then
        slowMA = quotes.GetSmma(PeriodSlowMovingAverage).ToResults()
    end

    if(TypeFastMovingAverage.Simple_SMA == true) then
        fastMA = quotes.GetSma(PeriodFastMovingAverage).ToResults()
    elseif(TypeFastMovingAverage.Exponential_EMA == true) then
        fastMA = quotes.GetEma(PeriodFastMovingAverage).ToResults()
    elseif(TypeFastMovingAverage.Weighted_WMA == true) then
        fastMA = quotes.GetWma(PeriodFastMovingAverage).ToResults()
    elseif(TypeFastMovingAverage.Smoothed_SMMA == true) then
        fastMA = quotes.GetSmma(PeriodFastMovingAverage).ToResults()
    end    

    
    local counter = #quotes
    for indexCandle = 0, counter -1 do
        if(slowMA[indexCandle].Item2 < fastMA[indexCandle].Item2) then
            outputs[indexCandle + 1] = (true == TriggerType.Bullish)
        else
            outputs[indexCandle + 1] = (false == TriggerType.Bullish)
        end

        if(indexCandle > 0) then
            if((slowMA[indexCandle - 1].Item2 < fastMA[indexCandle - 1].Item2 and TriggerType.Bullish and SignalType.CrossOver) or (slowMA[indexCandle - 1].Item2 > fastMA[indexCandle - 1].Item2 and TriggerType.Bearish and SignalType.CrossOver)) then
                outputs[indexCandle + 1] = false
            end
        end

        chartFMA[indexCandle + 1] =  {fastMA[indexCandle].Item1, fastMA[indexCandle].Item2}
        chartSMA[indexCandle + 1] = {slowMA[indexCandle].Item1, slowMA[indexCandle].Item2}
    end
    this.AddLineOnTheChart(chartSMA,"SMA","#FF0000",4)
    this.AddLineOnTheChart(chartFMA,"FMA")
    return outputs
end
```


Each user can backtest his own strategy and publish his strategy to run on Binance or Bybit.

**Reply by DaveSkender** [reply]
This is all very cool.  Thanks for sharing.
